index.bs

```diff
@@ -645,6 +648,26 @@ Modify the <a spec=HTML>download the hyperlink</a> algorithm to ensure that down
 
 Many specifications need to be patched so that, if a given algorithm invoked in a [=prerendering browsing context=], most of its work is deferred until the browsing context is [=prerendering browsing context/activated=]. This is tricky to do uniformly, as many of these specifications do not have great hygeine around <a href="https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-for-spec-authors">using the event loop</a>. Nevertheless, the following sections give our best attempt.
 
+<h4 id="delay-while-prerendering">The {{[DelayWhilePrerendering]}} extended attribute</h4>
+
+To abstract away some of the boilerplate involved in delaying the action of asynchronous methods until [=prerendering browsing context/activated|activation=], we introduce the <dfn extended-attribute>[DelayWhilePrerendering]</dfn> Web IDL extended attribute. It indicates that when a given method is called in a [=prerendering browsing context=], it will immediately return a pending promise and do nothing else. Only upon activation will the usual method steps take place, with their result being used to resolve or reject the previously-returned promise.
+
+The {{[DelayWhilePrerendering]}} extended attribute must take no arguments, and must only appear on a <a lt="regular operation" spec="WEBIDL">regular</a> or <a spec="WEBIDL">static operation</a> whose <a spec="WEBIDL">return type</a> is a <a spec="WEBIDL">promise type</a> and whose <a spec="WEBIDL">exposure set</a> contains {{Window}}.
+
+<div algorithm="DelayWhilePrerendering method steps">
+  The method steps for any operation annotated with the {{[DelayWhilePrerendering]}} extended attribute are replaced with the following:
+
+  1. Let |realm| be the [=current Realm=].
+  1. If the operation in question is a <a spec="WEBIDL">regular operation</a>, then set |realm| to the [=relevant Realm=] of [=this=].
+  1. If |realm|'s [=Realm/global object=]'s [=Window/browsing context=] is a [=prerendering browsing context=], then:
+    1. Let |promise| be [=a new promise=], created in |realm|.
+    1. Append the following steps to [=this=]'s [=platform object/post-prerendering activation steps list=]:
```

Comment:
What do you think about replacing the "post-prerendering activation steps list" construct with the machinery in https://heycam.github.io/webidl/#es-promise-manipulation?

Then we could use the WebIDL concept of reacting to a promise here instead:

```
1. Let |preconditions| be an empty [=list=].
1. If |realm|'s ... is a [=prerendering browsing context=], then:
  1. Append [=this=]'s [=prerendering promise=] to |preconditions|.
1. If |preconditions| is [=empty=], run the originally-specified steps for this operation, with the same [=this=] and arguments, and return the result.
1. [=Upon fulfillment=] of |preconditionPromise|, run the originally-specified steps for this operation, with the same [=this=] and arguments, and return the result.
```

(The silly step for the no preconditions case is probably necessary so that the immediate steps don't get delayed by until the microtask checkpoint.)

I personally like this construct better than data structures that store lists of steps, but your preference may vary, or I may be overthinking.

Reply:
I guess in your example `|preconditions|` is meant to be `|preconditionPromise|` and not be a list, but instead a promise-or-null or something like that? It seems reasonable, although I guess personally I get worried about using non-developer-exposed promises, as promises are _so_ JavaScriptey (e.g., realm specific) so they can be awkward to use as a spec-internal mechanism.